File: 653-两数之和IV-输入BST.py
from typing import Optional
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        # dfs+双指针
        # def dfs(cur: TreeNode):
        #     if not cur:
        #         return
        #     dfs(cur.left)
        #     nums.append(cur.val)
        #     dfs(cur.right)

        # nums = []
        # dfs(root)
        # left, right = 0, len(nums) - 1
        # while left < right:
        #     total = nums[left]+nums[right]
        #     if total > k:
        #         right -= 1
        #     elif total < k:
        #         left += 1
        #     else:
        #         return True
        # return False
        # 迭代+哈希
        hashset = set()
        que = deque([root])
        while que:
            cur = que.popleft(que)
            if k-cur.val in hashset:
                return True
            hashset.add(cur.val)
            if cur.left:
                que.append(cur.left)
            if cur.right:
                que.append(cur.right)
        return False

File: 653-两数之和IV-输入BST.cpp
#include <iostream>
using namespace std;
#include <string>
#include <unordered_set>
#include <deque>
#include <stack>
#include <queue>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution
{
public:
    unordered_set<int> hashset;
    bool findTarget(TreeNode *root, int k)
    {
        // DFS+哈希
        // if (!root)
        //     return false;
        // if (hashset.count(k - root->val))
        //     return true;
        // hashset.insert(root->val);
        // return findTarget(root->left, k) || findTarget(root->right, k);
        // BFS+哈希
        queue<TreeNode *> que;
        que.push(root);
        while (!que.empty())
        {
            TreeNode *cur = que.front();
            que.pop();
            if (hashset.count(k - cur->val))
                return true;
            hashset.insert(cur->val);
            if (cur->left)
                que.push(cur->left);
            if (cur->right)
                que.push(cur->right);
        }
        return false;
    }
};